### ISSUE TITLE
Working with High Volatility Stock Series

### ISSUE THREAD

**Issue by GabrielDeza** [awslabs/gluon-ts#935]
Hi,

For an academic project, I am trying to predict a companies opening stock, given the tweets related to the company between the previous day market close and the current day's market open. The goal at the end would be to attack this model in different ways and then come up with some defences to these well known attacks.

So far, I have a full pipeline of collecting tweets and stock prices for a company, resulting into a couple time series: Change in Stock, Average Sentiment, Number of tweets, and then a couple other time series such as average likes, average replies, etc. The issue I am facing lies in the incredibly volatile nature of change in stocks. I tried GPVar and DeepVar, and can sometimes get the right direction, but struggle immensely with the magnitude of the prediction.
![GPVAR](https://user-images.githubusercontent.com/53381889/87853379-5eaad800-c8d7-11ea-8e36-71a169f55adc.png)
Another problem I am facing when training GPVar or DeepVar for more then 20 epochs is just predicting pretty much the same value +/- some noise. The value seems in line with the median value in the training period. The following is a zoomed on the prediction period.
![GPVAR_Zoomed_gru](https://user-images.githubusercontent.com/53381889/87853441-0b855500-c8d8-11ea-8857-743f89d1fad3.png)

I was wondering if there are techniques to overcome highly volatile data or avoiding this local minimum problem of predicting a constant value. If you have any questions regarding my implementation, any questions or advice are greatly appreciated.

Thank you!

**Comment by StatMixedML** [comment]
@GabrielDeza Can you be more specific on the following:
 - What is your time frequency, i.e., weekly, daily, hourly? If it is non-high-frequency data (i.e., tick-data, seconds, minutes,...) with very long history, you wouldn't want to use Deep-Learning, but maybe, e.g., a multivariate ARMA-GARCH would do? Using ~ 1 year of daily data, as indicated by the plot, is far too little data.
 - Can you start 'simple' with a univariate model, like DeepAR, with a properly chosen distribution or Quantile Splines to see if it changes the picture.

**Comment by antonkiselevtdg** [comment]
Interesting.

I have tried predicting volatile currency pairs - I was surprised that univariate models performed much better than multivariate GPvar. Frequency was days.

I am currently testing combinations of univariate models, so far they are showing more promising results.

**Comment by GabrielDeza** [comment]
Hi StatMixedML and atonkiselevtdg,

I am using daily data (skip weekends since market is not open). In the plot above I had ~400 days worth of data. Since then I expanded my data set to 4 years to give me ~1000 data points. Is there a certain lower bound of data that would merit using deep-learning models for my problem? 
I am relatively knew to time series as well as this library so I apologize in advance. I tried a simple univariate model of DeepAR with the default student-t distribution, shown below. The parameters are shown below:
`prediction_length = 195,
context_length = 195,
freq = '1B',
epochs =40,
learning rate = 1e-3,
batch size =32,
num_batches_per_epoch =20`
<img width="1042" alt="Screen Shot 2020-07-20 at 3 00 12 PM" src="https://user-images.githubusercontent.com/53381889/87975716-fc4e0500-ca99-11ea-82cb-7741e9e9b845.png">

One thing that I am confused about is how DeepAR learns. When I only pass in 1 time series (change in stock), I receive a different loss if I pass in 2 or more time series. From my understanding, both of these time series are being learnt, so the loss represents them both. But in my case, I would like to have a M:1 model, where I only predict 1 time series given M time series. I thought that is what GPVAR and DeepVAR do but I think they might be M:M (ie: all time series are predicted). Hence I am worried that when I am using GPVAR and DeepVAR and decreasing loss, I am only decreasing the loss of my other times series and not the loss of the time series of interest (change in stock).

**Comment by GabrielDeza** [comment]
@antonkiselevtdg, just to make sure I understand, how I see prediction length of 195, is a test set of 195 data points, and the rest of the data (ie: everything before) is a training set. Is this correct?

**Comment by mbohlkeschneider** [comment]
Two comments on this, one more general and one more specific to GPVAR. 

1. For this kind of time series which are basically the same mean value with some random fluctuations without clear seasonality, forecasting the mean value is actually reasonable. If the model would try to predict the spikes but misses them, the error/loss would be higher. 

2. GPVAR uses a marginal transformation to make the data more Gaussian. Part of this transformation is an empirical CDF, which clips the extreme values to reduce the variance. Thus, the GPVAR cannot model these extreme values well, which might make it not a great choice for this kind of "spikey" data, even if it is multivariate. 

Hope that helps.

**Comment by antonkiselevtdg** [comment]
> @antonkiselevtdg, just to make sure I understand, how I see prediction length of 195, is a test set of 195 data points, and the rest of the data (ie: everything before) is a training set. Is this correct?

For DeepAR
http://gluon-ts.mxnet.io/api/gluonts/gluonts.model.deepar.html

prediction_length – Length of the prediction horizon

context_length – Number of steps to unroll the RNN for before computing predictions (default: None, in which case context_length = prediction_length)

**Comment by GabrielDeza** [comment]
Thank you mbohlkeschneider for the comments, point 1 makes sense to me (I though of it like a local minimum), I just wonder if there is a way to get around it. I didn't know that about GPVar, I will read the paper more in depth and try out DeepVar instead.
> prediction_length = 195 days is HUGE for volatile markets imho. I usually try to predict a specific value within 1-3 days or weeks, because of volatility and role of recent market changes. It is totally ok that it looks almost as a straight line with 195 days horizon.

antonkiselevtdg, thank you for your help, it is very much appreciated after a week of getting nowhere with this problem.
I am confused how the length of the prediction horizon affects my model. In a traditional ML problem such as a classification task, a sample from our test set denoted "example X", will have the same output if example X was the first example in our test set or the last. In the same sense, in my prediction horizon, I have all values of the other time series (ie: the inputs) and would just like to predict the singular output  (change in stock). Is this analogy not valid for time series?

**Comment by antonkiselevtdg** [comment]
> Thank you mbohlkeschneider for the comments, point 1 makes sense to me (I though of it like a local minimum), I just wonder if there is a way to get around it. I didn't know that about GPVar, I will read the paper more in depth and try out DeepVar instead.
> 
> prediction_length = 195 days is HUGE for volatile markets imho. I usually try to predict a specific value within 1-3 days or weeks, because of volatility and role of recent market changes. It is totally ok that it looks almost as a straight line with 195 days horizon.
> 
> antonkiselevtdg, thank you for your help, it is very much appreciated after a week of getting nowhere with this problem.
> I am confused how the length of the prediction horizon affects my model. In a traditional ML problem such as a classification task, a sample from our test set denoted "example X", will have the same output if example X was the first example in our test set or the last. In the same sense, in my prediction horizon, I have all values of the other time series (ie: the inputs) and would just like to predict the singular output  (change in stock). Is this analogy not valid for time series?

Length of prediction is a HUGE thing. The further your horizon (number of steps), the more difficult it is to predict a specific value, and the more prediction would look like a straight line (usually close to average value). Good article: https://blogs.oracle.com/datascience/7-ways-time-series-forecasting-differs-from-machine-learning

**Comment by mbohlkeschneider** [comment]
> Thank you mbohlkeschneider for the comments, point 1 makes sense to me (I though of it like a local minimum), I just wonder if there is a way to get around it. I didn't know that about GPVar, I will read the paper more in depth and try out DeepVar instead.
> 

Maybe if you have indicator features that mark whether a spike is happening. From the time series alone, I doubt that this will be possible. But if you find a way, I would be happy to learn :-).

**Comment by GabrielDeza** [comment]
@antonkiselevtdg Thank you for the information. I now understand how prediction length is not the size of my test/validation set, it is how much data I am looking to forecast in the future (horizon). Then my only remaining concern would how to determine if my model/parameters are any good. If the idea of a test set is still applicable to time series, I thought of the following but I am unsure if this works.

if I have N data points and a prediction length of 3 and would like to train on ~80% of my data, and test on ~20% of my data (where 0.2 * N >> 3). I would test on subsections of my test dataset of 3 long and determine some sort of metric from it. Then repeat this to cover the whole test set to get a global accuracy/loss, like the following diagram. Is this an appropriate method for determining generalizability of my time series model?
<img width="1006" alt="Screen Shot 2020-07-22 at 12 32 03 PM" src="https://user-images.githubusercontent.com/53381889/88203576-0ac22b00-cc18-11ea-8e01-4f6d866a9a9b.png">

**Comment by antonkiselevtdg** [comment]
> @antonkiselevtdg Thank you for the information. I now understand how prediction length is not the size of my test/validation set, it is how much data I am looking to forecast in the future (horizon). Then my only remaining concern would how to determine if my model/parameters are any good. If the idea of a test set is still applicable to time series, I thought of the following but I am unsure if this works.
> if I have N data points and a prediction length of 3 and would like to train on ~80% of my data, and test on ~20% of my data (where 0.2 * N >> 3). I would test on subsections of my test dataset of 3 long and determine some sort of metric from it. Then repeat this to cover the whole test set to get a global accuracy/loss, like the following diagram. Is this an appropriate method for determining generalizability of my time series model?

You can use rolling forecast. Your training set will "move in time" with your testing set.

Exanple:
Dataset May1 - June1
You want to use 10 days of data to forecast 1 next day

May 1-10 training set May 11 testing set
May 2-11 raining set May 12 testing set
...

https://github.com/awslabs/gluon-ts/issues/831

**Comment by GabrielDeza** [comment]
I see that people have worked on implementing on it but I am unsure if it has been implemented. If it hasn't I will write my own version.
I also think I finally understood the difference between adding a time series versus adding features. In my case, I only care about 1 time series, and all the others are features (sentiment of tweets, volume etc). However right now, I currently have no features, and am trying to predict all of them. I think this is wrong, so I need to convert them into features (like feat_dynamic_real etc). 
Other then that, I will close this thread within the day and I appreciate the help with my concerns :)
